packages/test/functional-tests/src/test/containerRuntime.spec.ts

```diff
@@ -241,5 +243,68 @@ describe("Container Runtime", () => {
             assert.strictEqual(2, batchBegin, "Did not receive correct batchBegin event for the batch");
             assert.strictEqual(2, batchEnd, "Did not receive correct batchEnd event for the batch");
         });
+
+        it("reconnects after receiving a leave op", async () => {
+            const service = new MockDocumentService(
+                undefined,
+                () => deltaConnection.disposed ? new MockDocumentDeltaConnection("test") : deltaConnection,
+            );
+
+            // overwriting connection mode to return the requested mode from client
+            // instead of default to "write" in MockDocumentDeltaConnection
+            service.connectToDeltaStream = async (newClient: IClient): Promise<IDocumentDeltaConnection> => {
+                const connection = deltaConnection.disposed ? new MockDocumentDeltaConnection("test") : deltaConnection;
+                connection.mode = newClient.mode;
+                return connection;
+            };
+            const client = { mode: "write", details: { capabilities: { interactive: true } } };
+
+            const deltaManager2 = new DeltaManager<ConnectionManager>(
+                () => service,
+                DebugLogger.create("fluid:testDeltaManager"),
+                () => true,
+                (props: IConnectionManagerFactoryArgs) => new ConnectionManager(
+                    () => service,
+                    client as IClient,
+                    true,
+                    DebugLogger.create("fluid:testConnectionManager"),
+                    props),
+            );
+            await deltaManager2.attachOpHandler(0, 0, 1, {
+                process(message: ISequencedDocumentMessage) {
+                    processOp(message);
+                    return {};
+                },
+                processSignal() { },
+            });
+            await new Promise((resolve) => {
+                deltaManager2.on("connect", resolve);
+                deltaManager2.connect({ reason: "test" });
+            });
+            assert(deltaManager2.active, "deltaManager should be active");
+            assert.strictEqual(deltaManager2.connectionManager.connectionMode, "write",
+                "connection mode should be write");
+
+            const leaveMessage: ISequencedDocumentSystemMessage = {
+                clientId: "null",
+                data: `"${deltaManager2.connectionManager.clientId}"`,
+                minimumSequenceNumber: 0,
+                sequenceNumber: seq++,
+                type: MessageType.ClientLeave,
+                term: 1,
+                clientSequenceNumber: 1,
+                referenceSequenceNumber: 1,
+                contents: "",
+                timestamp: 1,
+            };
+
+            deltaConnection.emitOp(docId, [leaveMessage]);
+            // Yield the event loop because the inbound op will be processed asynchronously.
+            await yieldEventLoop();
```

Comment:
You might be getting lucky here with yieldEventLoop() - not sure. I'd rather wait again (same way as you wait above) for a new "connect" event instead - it's more telling on what's happening.